samples/object_api.c

```diff
@@ -0,0 +1,338 @@
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/**
+ * @file object_api.c
+ * @brief A code sample illustrates the basic usage of the OPAE C Object API.
+ *
+ *
+ */
+#include <opae/fpga.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <getopt.h>
+#include <string.h>
+#include <unistd.h>
+
+void print_err(const char *s, fpga_result res)
+{
+	fprintf(stderr, "Error %s: %s\n", s, fpgaErrStr(res));
+}
+
+#define ON_ERR_GOTO(res, label, desc)                                          \
+	do {                                                                   \
+		if ((res) != FPGA_OK) {                                        \
+			print_err((desc), (res));                              \
+			goto label;                                            \
+		}                                                              \
+	} while (0)
+
+
+typedef fpga_result (*destroy_f)(void *p);
+#define MAX_CLEANUP 100
+struct {
+	destroy_f fn;
+	void *param;
+} cleanup[MAX_CLEANUP];
+
+static int cleanup_size = 0;
+
+#define ADD_TO_CLEANUP(func, p)                                                \
+	if (cleanup_size < MAX_CLEANUP) {                                      \
+		cleanup[cleanup_size].fn = (destroy_f)func;                    \
+		cleanup[cleanup_size++].param = p;                             \
+	}
+#define MAX_TOKENS 4
+#define MAX_GROUP_OBJECTS 32
+typedef struct {
+	const char *name;
+	fpga_object object;
+	uint64_t value;
+	uint64_t delta;
+} named_object;
+
+typedef struct {
+	const char *name;
+	fpga_token token;
+	fpga_object object;
+	uint8_t bus;
+        uint8_t device;
+        uint8_t function;
+	named_object objects[MAX_GROUP_OBJECTS];
+	size_t count;
+} metric_group;
+
+typedef struct {
+	fpga_token token;
+	metric_group *groups;
+	size_t count;
+	fpga_object clock;
+} token_group;
+
+struct {
+	bool freeze;
+	int bus;
+	float interval_sec;
+} options = { false, -1, 1.0 };
```

Comment:
What is this options about? When freeze is set, what the bus and interval_sec values mean?

Reply:
Freeze is used to tell the sample to freeze (take a snapshot) of the counters before reading them.
Writing to the "freeze" sysfs file tells the driver to snapshot all counters in the sysfs directory.

bus is used to filter when enumerating. interval_usec is the microseconds to sleep in between reading perf counters.